File: backend/src/stats.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

#[derive(Debug, Clone, serde::Serialize)]
pub struct RealtimeStats {
    pub timestamp: u64,
    pub total_speed_bps: u64,
    pub per_host_speed: HashMap<String, u64>,
}

#[derive(Debug, Clone)]
pub struct StatsManager {
    ring_buffer: Arc<RwLock<Vec<RealtimeStats>>>,
    buffer_size: usize,
}

impl StatsManager {
    pub fn new(buffer_size: usize) -> Self {
        Self {
            ring_buffer: Arc::new(RwLock::new(Vec::with_capacity(buffer_size))),
            buffer_size,
        }
    }

    pub async fn record_stats(&self, total_speed_bps: u64, per_host_speed: HashMap<String, u64>) {
        let timestamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();

        let stats = RealtimeStats {
            timestamp,
            total_speed_bps,
            per_host_speed,
        };

        let mut buffer = self.ring_buffer.write().await;
        buffer.push(stats);
        if buffer.len() > self.buffer_size {
            buffer.remove(0);
        }
    }

    pub async fn get_realtime_stats(&self) -> Vec<RealtimeStats> {
        self.ring_buffer.read().await.clone()
    }

    pub async fn get_current_stats(&self) -> Option<RealtimeStats> {
        self.ring_buffer.read().await.last().cloned()
    }
}

File: backend-go/internal/stats/stats.go
package stats

import (
	"sync"
	"time"
)

// Portado de backend/src/stats.rs:5 — mantém mesma forma JSON (snake_case)
type RealtimeStats struct {
	Timestamp      uint64            `json:"timestamp"`
	TotalSpeedBps  uint64            `json:"total_speed_bps"`
	PerHostSpeed   map[string]uint64 `json:"per_host_speed"`
}

// StatsManager — ring buffer thread-safe, portado de stats.rs:13
type StatsManager struct {
	mu         sync.RWMutex
	buffer     []RealtimeStats
	bufferSize int
}

func New(bufferSize int) *StatsManager {
	if bufferSize <= 0 {
		bufferSize = 60
	}
	return &StatsManager{
		buffer:     make([]RealtimeStats, 0, bufferSize),
		bufferSize: bufferSize,
	}
}

func (m *StatsManager) RecordStats(totalSpeedBps uint64, perHostSpeed map[string]uint64) {
	if perHostSpeed == nil {
		perHostSpeed = map[string]uint64{}
	}
	// Clone map to avoid external mutation
	clone := make(map[string]uint64, len(perHostSpeed))
	for k, v := range perHostSpeed {
		clone[k] = v
	}
	timestamp := uint64(time.Now().Unix())
	s := RealtimeStats{
		Timestamp:     timestamp,
		TotalSpeedBps: totalSpeedBps,
		PerHostSpeed:  clone,
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	m.buffer = append(m.buffer, s)
	if len(m.buffer) > m.bufferSize {
		// remove oldest (ring behavior)
		copy(m.buffer[0:], m.buffer[1:])
		m.buffer = m.buffer[:m.bufferSize]
	}
}

func (m *StatsManager) GetRealtimeStats() []RealtimeStats {
	m.mu.RLock()
	defer m.mu.RUnlock()
	out := make([]RealtimeStats, len(m.buffer))
	copy(out, m.buffer)
	return out
}

func (m *StatsManager) GetCurrentStats() *RealtimeStats {
	m.mu.RLock()
	defer m.mu.RUnlock()
	if len(m.buffer) == 0 {
		return nil
	}
	last := m.buffer[len(m.buffer)-1]
	// return copy
	c := last
	return &c
}

// Global singleton usado pelo handler e pelo sampler em main.go
var DefaultManager = New(60)
